Auto-detection of Azure cloud from azure.json

johnharris85 opened the pull request:
Not tested this yet, will update once I have if necessary. 

Fairly minor change but should enable the service to auto-detect the cloud and resource manager endpoint directly to support China / Gov etc... There may be other things that don't work there :p But this will at least auto-detect. Uses the `cloud` attribute from the `azure.json` that's mounted into the ip-allocator daemonset.

johnharris85 commented:
`cloud` is a required field in the `/etc/kubernetes/azure.json` (same as the other auth fields) so should always be available. WDYT about just documenting it as a required field? And maybe making the error more explicit ("Have you checked your `cloud` field is set to a valid value?") or something? Falling back on defaults sometimes gets kind of hard to debug / figure out what's going on if it happens silently.

ddebroy commented:
I would prefer to keep compatibility with the existing behavior and not break it right away. Perhaps you can add a warning saying "Defaulting to Public Cloud as no `cloud` field is specified" and that the default will be removed in the future. This way, we can have a period of time where we keep compatibility with existing behavior while warning that the default will be removed in the future and ultimately make it mandatory in a few months.

johnharris85 commented:
Added the default behavior back in, and made it possible to override both methods via environment variable. Also added logging so it's clear where that value is coming from.
